### PR TITLE
feat(bonding-curve): implement bonding curve contract

### PR DESCRIPTION
**Pull request by yushih**

### CODE REVIEW

**Comment by yushih** [comment]
The `deposit` action requires user to transfer both PNK and ETH into the bonding contract. To my knowledge there is no way of doing this using only one transaction. Tthis may possibly be achieved by letting `approveAndCall` of the token contract take an ETH value and forward it to `receiveApproval`. However it requires updating the token contract. Is it safe and worth doing?. 

The method used in this PR is to require the depositor to first grant the bonding curve contact an amount of PNK on the token contract, by calling `approve`. Then the depositor should call `deposit` of the bonding curve contact with an amount of ETH to finish the deposition. 

An alternative approach is: 1. the depositor calls a method of the bonding curve, paying ETH. The contract records the payment. 2. the depositor calls `approveAndCall` of the token contract, which in turn calls `receiveApproval` of the bonding curve contract. Now `receiveApproval` has both the ETH and PNK and handles the deposition. This approach seems to be slightly more complex and has no advantage so it it not adopted.

**Comment by yushih** [comment]
Why the dependency on `solidity-bytes-utils`: 

To be able to sell PNK to the bonding curve contract, there must be a way of transferring PNK to the bonding curve contract on the token (PNK) contract.  This is implemented by requiring the seller to call `approveAndCall` on the token contract, which invokes `receiveApproval` of the bonding curve contract. We want to pass multiple parameters through to `receiveApproval` via `bytes _extraData` parameter of `approveAndCall`(https://github.com/Giveth/minime/blob/master/contracts/MiniMeToken.sol#L254) so these parameters must be packed into `bytes` and unpacked by `receiveApproval`. The unpacking (https://github.com/kleros/kleros-interaction/pull/146/commits/321d2bda13ee9dc2834cf857bf6ae227266b8270#diff-8bc92b5b91830fb57bae888bffd40a69R160) leverages `toAddress` and `toUint` from https://github.com/GNSPS/solidity-bytes-utils/blob/master/contracts/BytesLib.sol.  Those functions are simple and look correct to me. They are under public domain.

File: contracts/standard/bonding-curve/BondingCurve.sol [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+        // solium-disable-next-line indentation
+        require(_extraData[0]==0x62 && // 'b'
+            _extraData[1]==0x63 &&     // 'c'
+            _extraData[2]==0x73 &&     // 's'
+            _extraData[3]==0x31,       // '1'
+            "Expect magic number.");
+
+        address recipient = BytesLib.toAddress(_extraData, 4);
+        uint minEth = BytesLib.toUint(_extraData, 24);
```
Comment (epiqueras) [inline contracts/standard/bonding-curve/BondingCurve.sol]: ETH

File: contracts/standard/bonding-curve/BondingCurve.sol [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+        // Calculate the amount of ETH that should be paid to seller:
+        // To maintain (totalEth - eth)*(totalPnk+_amount) == totalEth*totalPnk
+        // we get eth = totalEth * _amount / (totalPnk + _amount)
```
Comment (epiqueras) [inline contracts/standard/bonding-curve/BondingCurve.sol]: to seller to maintain... . We get

File: contracts/standard/bonding-curve/BondingCurve.sol [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        // Then charge a spread. 
+        uint eth = totalEth.mul(_amount).mul(SPREAD_DIVISOR)
+            .div(totalPnk.add(_amount)).div(SPREAD_DIVISOR.add(spread));
+
+        require(eth >= minEth, "PNK price must be above minimum expected value.");
+        recipient.transfer(eth);
+        require(pinakion.transferFrom(_from, this, _amount), "PNK transfer failed.");
```
Comment (epiqueras) [inline contracts/standard/bonding-curve/BondingCurve.sol]: Possible reentrancy if malicious token.

File: contracts/standard/bonding-curve/BondingCurve.sol [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    }
+
+    // **************************** //
+    // *     Governor Functions   * //
+    // **************************** //
+
+    /** @dev Setter for spread.
+      * @param _spread Spread.
+      */
+    function setSpread(uint _spread) public onlyGovernor {
```
Comment (epiqueras) [inline contracts/standard/bonding-curve/BondingCurve.sol]: external

File: contracts/standard/bonding-curve/BondingCurve.sol [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        spread = _spread;
+    }
+
+    /** @dev Setter for governor.
+     *  @param _governor Address of the governor contract.
+     */
+    function setGovernor(address _governor) public onlyGovernor {
```
Comment (epiqueras) [inline contracts/standard/bonding-curve/BondingCurve.sol]: external

File: contracts/standard/bonding-curve/BondingCurve.sol [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+    /** @dev Setter for spread.
+      * @param _spread Spread.
+      */
+    function setSpread(uint _spread) public onlyGovernor {
+        spread = _spread;
+    }
+
+    /** @dev Setter for governor.
+     *  @param _governor Address of the governor contract.
```
Comment (epiqueras) [inline contracts/standard/bonding-curve/BondingCurve.sol]: The address

**Pull request merged**